core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -1106,6 +1115,12 @@ class Partition(val topicPartition: TopicPartition,
    * @return true if the HW was incremented, and false otherwise.
    */
   private def maybeIncrementLeaderHW(leaderLog: UnifiedLog, currentTimeMs: Long = time.milliseconds): Boolean = {
+    if (metadataCache.isInstanceOf[KRaftMetadataCache] && interBrokerProtocolVersion.isElrSupported && eligibleLeaderReplicasEnabled) {
```

Comment:
Do we need the check on the metadata cache here? Shouldn't it be impossible to be running in ZK mode with ELR enabled?

Reply:
I guess ZK can also upgrade to 3.7IV1 and the elrEnabled is a static config shared between Kraft and ZK?

Reply:
I have a higher-level question -- do we actually care if ELR is enabled for this logic?  It is my understanding that the logic on the brokers doesn't use ELR, it just has some changes that seem to improve the protocol even if ELR is not used / implemented:
- strict min ISR
- advance HWM only only if min ISR known to the controller

Reply:
Agree. Now make it a general requirement.